autopilot/contractor.go

```diff
@@ -221,6 +227,91 @@ func (c *contractor) performContractMaintenance(cfg api.AutopilotConfig, cs api.
 	return c.ap.bus.SetContractSet("autopilot", contractset)
 }
 
+func (c *contractor) performWalletMaintenance(cfg api.AutopilotConfig, cs api.ConsensusState) error {
+	c.logger.Info("performing wallet maintenance")
+	b := c.ap.bus
+	l := c.logger
+
+	// no contracts - nothing to do
+	numHostsWanted := cfg.Contracts.Hosts
+	if numHostsWanted == 0 {
+		l.Debug("wallet maintenance skipped, no contracts wanted")
+		return nil
+	}
+
+	// no allowance - nothing to do
+	if cfg.Contracts.Allowance.IsZero() {
+		l.Debug("wallet maintenance skipped, no allowance set")
+		return nil
+	}
+
+	// no money - nothing to do
+	balance, err := b.WalletBalance()
+	if err != nil {
+		return err
+	}
+	if balance.IsZero() {
+		l.Debug("wallet maintenance skipped, zero balance in wallet")
+		return nil
+	}
+
+	// balance below allowance - nothing to do
+	if balance.Cmp(cfg.Contracts.Allowance) < 0 {
+		l.Debugf("wallet maintenance skipped, balance lower than allowance %v<%v", balance, cfg.Contracts.Allowance)
+		return nil
+	}
+
+	// fetch pending transactions
+	pending, err := c.ap.bus.WalletPending()
+	if err != nil {
+		return nil
+	}
+
+	// pending maintenance transaction - nothing to do
+	c.mu.Lock()
+	currID := c.maintenanceTxnID
+	c.mu.Unlock()
+	for _, txn := range pending {
+		if currID == txn.ID() {
+			l.Debugf("wallet maintenance skipped, pending transaction found with id %v", currID)
+			return nil
+		}
+	}
+
+	// fetch wallet outputs
+	outputs, err := b.WalletOutputs()
+	if err != nil {
+		return err
+	}
+	numOutputs := uint64(len(outputs))
+
+	// enough outputs - nothing to do
+	if numOutputs >= numHostsWanted {
+		l.Debugf("no wallet maintenance needed, plenty of outputs available (%v>=%v)", numOutputs, numHostsWanted)
+		return nil
+	}
+
+	// check amount
+	amount := balance.Div64(numOutputs)
+	min, _ := initialContractFundingMinMax(cfg)
+	if amount.Cmp(min) < 0 {
+		amount = min.Mul64(2)
```

Comment:
Does this even work? We divide 100% of the balance by let's say 50. So if `amount < min` doesn't that mean we simply don't have enough money? Because if `min > amount` then `min * 50 > balance`

So maybe in the check above we should not compare the balance to the allowance but 50*minFunding to the allowance.

Also multiplying `min` by 2 works since the max is double min anyway but if we ever choose different divisors and are not careful about updating this too, we might end up checking against a value that is bigger than max.

Reply:
we can't really use `minFunding` either because of added contract price and fees, I'll use the average